module/bdev/lvol/vbdev_lvol.c

```diff
@@ -222,66 +222,9 @@ int
 vbdev_lvs_create(const char *base_bdev_name, const char *name, uint32_t cluster_sz,
 		 enum lvs_clear_method clear_method, spdk_lvs_op_with_handle_complete cb_fn, void *cb_arg)
 {
-	struct spdk_bs_dev *bs_dev;
-	struct spdk_lvs_with_handle_req *lvs_req;
-	struct spdk_lvs_opts opts;
-	int rc;
-	int len;
-
-	if (base_bdev_name == NULL) {
-		SPDK_ERRLOG("missing base_bdev_name param\n");
-		return -EINVAL;
-	}
-
-	spdk_lvs_opts_init(&opts);
-	if (cluster_sz != 0) {
-		opts.cluster_sz = cluster_sz;
-	}
-
-	if (clear_method != 0) {
-		opts.clear_method = clear_method;
-	}
-
-	if (name == NULL) {
-		SPDK_ERRLOG("missing name param\n");
-		return -EINVAL;
-	}
-
-	len = strnlen(name, SPDK_LVS_NAME_MAX);
-
-	if (len == 0 || len == SPDK_LVS_NAME_MAX) {
-		SPDK_ERRLOG("name must be between 1 and %d characters\n", SPDK_LVS_NAME_MAX - 1);
-		return -EINVAL;
-	}
-	snprintf(opts.name, sizeof(opts.name), "%s", name);
-
-	lvs_req = calloc(1, sizeof(*lvs_req));
-	if (!lvs_req) {
-		SPDK_ERRLOG("Cannot alloc memory for vbdev lvol store request pointer\n");
-		return -ENOMEM;
-	}
-
-	rc = spdk_bdev_create_bs_dev_ext(base_bdev_name, vbdev_lvs_base_bdev_event_cb,
-					 NULL, &bs_dev);
-	if (rc < 0) {
-		SPDK_ERRLOG("Cannot create blobstore device\n");
-		free(lvs_req);
-		return rc;
-	}
-
-	lvs_req->bs_dev = bs_dev;
-	lvs_req->base_bdev = bs_dev->get_base_bdev(bs_dev);
-	lvs_req->cb_fn = cb_fn;
-	lvs_req->cb_arg = cb_arg;
-
-	rc = spdk_lvs_init(bs_dev, &opts, _vbdev_lvs_create_cb, lvs_req);
-	if (rc < 0) {
-		free(lvs_req);
-		bs_dev->destroy(bs_dev);
-		return rc;
-	}
-
-	return 0;
+	/* set uuid as NULL */
+	return vbdev_lvs_create_with_uuid(base_bdev_name, name, NULL, cluster_sz, clear_method, cb_fn,
+					  cb_arg);
```

Comment:
While this removes much of the code duplication there is still a large diff effectively moving the code elsewhere and will make upstream changes more difficult to merge in future. Rename the existing `vbdev_lvs_create()` to `vbdev_lvs_create_with_uuid()`, insert the code related to `uuid` there, and re-introduce the former function to call through to the latter.

Reply:
`vbdev_lvs_create()` cannot be called from `vbdev_lvs_create_with_uuid()` as the struct `spdk_lvs_opts` will be re initialized inside `vbdev_lvs_create()` and the uuid will be lost. I did not want to change the function signature of the existing function so IMO having a duplicate makes more sense here.

Reply:
No, call `vbdev_lvs_create_with_uuid()` from `vbdev_lvs_create()`, exactly as it is currently implemented but without moving the code around.

Reply:
That would required adding `uuid` to `vbdev_lvs_create()` right?

Reply:
Or changing the argument of `vbdev_lvs_create()` to accept `spdk_lvs_opts` instead of all different fields.

Reply:
Basically this:
```
diff --git a/module/bdev/lvol/vbdev_lvol.c b/module/bdev/lvol/vbdev_lvol.c
index 6f2c0c976..b72d929eb 100644
--- a/module/bdev/lvol/vbdev_lvol.c
+++ b/module/bdev/lvol/vbdev_lvol.c
@@ -219,6 +219,15 @@ end:
 int
 vbdev_lvs_create(const char *base_bdev_name, const char *name, uint32_t cluster_sz,
                 enum lvs_clear_method clear_method, spdk_lvs_op_with_handle_complete cb_fn, void *cb_arg)
+{
+       return vbdev_lvs_create_with_uuid(base_bdev_name, name, NULL, cluster_sz, clear_method, cb_fn,
+                                         cb_arg);
+}
+
+int
+vbdev_lvs_create_with_uuid(const char *base_bdev_name, const char *name,  const char *uuid,
+                uint32_t cluster_sz,
+                enum lvs_clear_method clear_method, spdk_lvs_op_with_handle_complete cb_fn, void *cb_arg)
 {
        struct spdk_bs_dev *bs_dev;
        struct spdk_lvs_with_handle_req *lvs_req;
@@ -240,6 +249,10 @@ vbdev_lvs_create(const char *base_bdev_name, const char *name, uint32_t cluster_
                opts.clear_method = clear_method;
        }
 
+       if (uuid != NULL) {
+               opts.uuid = uuid;
+       }
+
        if (name == NULL) {
                SPDK_ERRLOG("missing name param\n");
                return -EINVAL;
```